lerobot/scripts/server/robot_client.py

```diff
@@ -0,0 +1,608 @@
+            send_end = time.time()
+
+            obs_timestep = obs.get_timestep()
+
+            self.logger.info(
+                f"Sent observation #{obs_timestep} | "
+                f"Serialize time: {serialize_time - start_time:.6f}s | "
+                f"Network time: {send_end - send_start:.6f}s | "
+                f"Total time: {send_end - start_time:.6f}s"
+            )
+
+            self.last_obs_sent_time = send_end
+            return True
+
+        except grpc.RpcError as e:
+            self.logger.error(f"Error sending observation #{obs.get_timestep()}: {e}")
+            return False
+
+    def _validate_action(self, action: TimedAction):
+        """Received actions are keps only when they have been produced for now or later, never before"""
+        return not action.get_timestep() <= self.latest_action
+
+    def _inspect_action_queue(self):
+        queue_size = self.action_queue.qsize()
+        timestamps = sorted([action.get_timestep() for action in self.action_queue.queue])
+        self.logger.debug(f"Queue size: {queue_size}, Queue contents: {timestamps}")
+        return queue_size, timestamps
+
+    def _update_action_queue(self, actions: list[TimedAction]):
+        """Update the action queue with new actions, without ever emptying the queue"""
+
+        new_queue = Queue()
+        for action in actions:
+            if self._validate_action(action):
+                new_queue.put(action)
+
+        self.action_queue = new_queue
+
+    def _aggregate_action_queues(
+        self,
+        incoming_actions: list[TimedAction],
+        aggregate_fn: Optional[Callable[[torch.Tensor, torch.Tensor], torch.Tensor]] = None,
+    ):
+        """Finds the same timestep actions in the queue and aggregates them using the aggregate_fn"""
+        # TODO(fracapuano): move outside of the function and make aggregate_fn an always required argument
+        if not aggregate_fn:
+            # default aggregate function: take the latest action
+            def aggregate_fn(x1, x2):
+                return x2
+
+        action_intersections: list[torch.Tensor] = []
+        current_action_queue = {
+            action.get_timestep(): action.get_action() for action in self.action_queue.queue
+        }
+
+        for new_action in incoming_actions:
+            if new_action.get_timestep() in current_action_queue:
+                # TODO(fracapuano): There is probably a way to do this with broadcasting of the two action tensors
+                action_intersections.append(
+                    TimedAction(
+                        timestamp=new_action.get_timestamp(),
+                        action=aggregate_fn(
+                            current_action_queue[new_action.get_timestep()], new_action.get_action()
+                        ),
+                        timestep=new_action.get_timestep(),
+                    )
+                )
+            else:
+                action_intersections.append(new_action)
```

Comment:
Nit: Simplify, most of the time else statement is not useful and adds cognitive overload to the reader due to the indentation, if we may avoid it.
```suggestion
            if new_action.get_timestep() not in current_action_queue:
                   action_intersections.append(new_action)
                   continue
                
                # TODO(fracapuano): There is probably a way to do this with broadcasting of the two action tensors
            current_action = current_action_queue[timestep]
            aggregated_action = aggregate_fn(current_action, new_action.get_action())

            action_intersections.append(
                TimedAction(
                    timestamp=new_action.get_timestamp(),
                    action=aggregated_action,
                    timestep=timestep,
                )
            )

```

I think you can simplify

Reply:
Closing this as it is outdated